### PR TITLE
HVPA Redesign Proposal

### PR DESCRIPTION
**Pull request by amshuman-kr**
**What this PR does / why we need it**:
Redesign proposal for HVPA to support more efficient and consistent horizontal and vertical scaling.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:
A rejigged version of #71. I have raised this as a separate proposal in the interest of keeping the discussion in the main repo rather than in a fork. The main file is [HVPA-EP-001.md](https://github.com/gardener/hvpa-controller/pull/74/files#diff-2a6ff5bc627629da1aa8e7fe6abfe150).

**Release note**:

```improvement operator
NONE
```

### CODE REVIEW

**Comment by vlerenc** [comment]
It is an interesting thought to make `baseResourcesPerReplica` configurable. However, it will increase complexity. In any case, is the formula really right: `(total resource recommendation} / desiredReplicas) + max(spec.baseResourcesPerReplica[resource].value, (total resource recommendation) x spec.baseResourcesPerReplica[resource].percentage / 100)`? Since VPA has indirectly seen the base usage already, don't we have to first subtract the current base usage first, i.e. introduce a term `- currentReplicas x baseUsage` (where `baseUsage` is a simplistic form of the actual configured base usage configuration omitted here for simplicity)?

**Comment by amshuman-kr** [comment]
@vlerenc Thanks for the review and feedback!

> 1. ... Since VPA has indirectly seen the base usage already, don't we have to first subtract the current base usage first, i.e. introduce a negative term such as - currentReplicas x baseUsage

Good catch. You are right. @hat I wrote would over budget for the base usage. I will rewrite that part to be more accurate.

> 2. ... While the document is open to it, would we really implement HPA+VPA recommendations?

I wanted to keep the approach/design such that it can work with any combination HPA or VPA or HPA+VPA or some other combination involving some other future sources or recommendations (predictive?). But I agree, HPA doesn't necessarily help here. I remember (@ggaurav10 correct me if I am wrong) we have see HPA also not react quickly enough for spikes in load (though, at the configured 80% average utilisation it is more generous in recommendation than VPA).

> 3. ... Which one will you go for and why? I think it is ScalingIntervalsOverlap from the later API description, but why was this approach chosen?

Yes. In the design section, I went with `ScalingIntervalsOverlap`. The only reason was the simplicity of API (it needs to be configured once in the spec).

I agree `ScaleDownMinResource` approach is more flexible as the overlap can be configured differently for different scaling intervals. However, it is not the full picture. The scaling intervals in the spec are not necessarily _effective scaling intervals_ for every replica (though, of course, they can be specified as such). So, if they are specified for ranges of _effective scaling intervals_ then we need some additional description (or additional API) of how to extrapolate (say, linearly)  `ScaleDownMinResource` for the computed _effective scaling intervals_. I am open to this though because it would be more flexible.

> 4. We have a practical problem (stated indirectly as non-goal, I believe) that VPA may make down-scale recommendations or inappropriate up-scale recommendations that would violate even the current usage (e.g. CPU). Any thoughts on incorporating that? Would we try to solve that via VPA (hard to change upstream) or a third recommender (besides HPA and VPA) that at least recommends the current usage (since we always pick the max from all recommendations)? That could be a trivial thing to avoid VPA running amok. WDYT?

My current thinking is to introduce a new recommender but not as an additional 3rd recommendation source to HPA but to replace VPA. I have high level idea (yet to finalise details) about what it should be. Ideally, it should be simple enough improvement over current VPA recommender. But considering the turn-around time for upstream contribution, I think we might have to do this on our own (either as a fork or from scratch). Also, I think it should be possible to deploy it alongside regular VPA recommender and see observe to see if it actually recommends better than VPA.

I am hesitating to introduce a check in HVPA regarding current usage because current usage is potentially a fleeting thing and HVPA might end up reacting too quickly to spikes if we implement it trivially. And if we implement it non-trivially, it is anyway better done as separate recommender. WDYT?

> 5. ... In addition, if and when we have it, wouldn't it also remove the need for HPA altogether?

Agreed. HPA seems not fit for most non-trivial use-cases. Certainly, as a way to apply the recommendations if not also as the source of recommendations. But only eventually, of course, once HVPA implementation is there.

**Comment by vlerenc** [comment]
> I wanted to keep the approach/design such that it can work with any combination HPA or VPA...

Yes, I like that @amshuman-kr .

> The only reason was the simplicity of API [with `ScalingIntervalsOverlap`] (it needs to be configured once in the spec).

OK, understood. It's the more general approach, also if we would have "computed" non-explicit buckets.

> My current thinking is to introduce a new recommender but not as an additional 3rd recommendation source to HPA but to replace VPA.

Ah, OK.

> I think it should be possible to deploy it alongside regular VPA recommender and see observe to see if it actually recommends better than VPA.

Nice.

> I am hesitating to introduce a check in HVPA regarding current usage because current usage is potentially a fleeting thing and HVPA might end up reacting too quickly to spikes if we implement it trivially.

You are right. My thought was poor/over-simplistic.

File: docs/proposals/HVPA-EP-0001/HVPA-EP-001.md [hunk 1]
```diff
@@ -0,0 +1,803 @@
...
+
+##### Maximum of the `desiredReplicas` based on the total resource recommendation for each of the containers
+
+Another alternative is to treat this conflict in the same way as the conflict between the recommendations for [multiple resources](#multiple-resources) for the same container and extend the same solution.
+I.e. use the higher value of `desiredReplicas` among those specified for all the resources of all the container in the pod based on the [total resource recommendation](#total-resource-recommendation) for that resource for that container.
+Once the `desiredReplicas` has been decided in this way, the adjusted resource recommendation per replica for each resource and each container would be calculated based on the [total resource recommendation](#total-resource-recommendation) for that resource for that container.
+
+This approach is recommended in favour of [redefining total resource recommendation](#sum-of-the-recommendations-for-a-resource-for-all-the-containers-as-total-recommendation-for-the-resource).
```
Comment (ggaurav10) [inline docs/proposals/HVPA-EP-0001/HVPA-EP-001.md:442]: As discussed earlier, same set of `scalingIntervals` or buckets will be used to calculate `desiredReplicas` and the adjusted resource recommendation per replica for all the containers in a pod.
Do we also use the same `BaseResourcesPerReplica` too for all the containers when adjusting for the base resource usage?

Comment (amshuman-kr) [reply docs/proposals/HVPA-EP-0001/HVPA-EP-001.md:442]: Yes, for the time-being. Unless the use-case for different configuration for different containers becomes strong.

**Review by ggaurav10 — APPROVED** [review]

**Comment by ggaurav10** [comment]
looks like https://github.com/gardener/hvpa-controller/pull/74#discussion_r449944689 didn't go in the final merged commit. Could be because it was directly merged from github web and local git branch wasn't rebased.
cc @amshuman-kr

**Comment by amshuman-kr** [comment]
> looks like [#74 (comment)](https://github.com/gardener/hvpa-controller/pull/74#discussion_r449944689) didn't go in the final merged commit. Could be because it was directly merged from github web and local git branch wasn't rebased.
> cc @amshuman-kr

Thanks for pointing out 👍 
I had rebased, not sure why. I will file a separate PR to fix.

**Pull request merged**